### ISSUE TITLE
Support for custom java.lang.reflect.Type

### ISSUE THREAD

**Issue by serj-lotutovici** [square/moshi#113]
Currently if someone creates a custom implementation of `java.lang.reflect.Type` (with a custom `JsonAdapter` respectively) the utility method `Types.getRawType()` will throw a `IllegalArgumentException` doe to support only for `Class<?>`, `ParameterizedType` e.t.c. A workaround for this would be to just use `type instanceof MyCustomType` in the adapter factory, but then the custom adapter has to be first in line, so that other adapters don't throw through `Types.getRawType()`.

This is ok till the point when users of an api that defines the custom type (which is fully private and is an implementation detail) add their own adapters.

By trying to implement the `ParameterizedType` puts the curtom adapter in conflict with Moshi's collection adapters, and doesn't work more sophisticated use cases like `MyCustomType[List<MyCustomType>]`.

**Comment by swankjesse** [comment]
Concrete example?

**Comment by serj-lotutovici** [comment]
We have a java unfriendly api. And unfortunately we can not request to change it, since a lot of teams already adopted these models. Most responses come in such a way:
```json
{
  "users": [ {}, {}, {}, ... ]
}
```
or something even crazy like:
```json
{
  "result": [
    {
       "user": {}
    },
   ...
  ]
```

To simplify things we declared a custom `Type` which takes the desired result type and a list of roots, where that object can be found. The adapter go's through the roots, and returns the final type.

You can find the code here:
- [Our custom type](https://github.com/xing/xing-android-sdk/blob/master/sdk/src/main/java/com/xing/android/sdk/CompositeType.java)
- [The adapter](https://github.com/xing/xing-android-sdk/blob/master/sdk/src/main/java/com/xing/android/sdk/CompositeTypeJsonAdapter.java)
- [Tests to show the desired behaviour](https://github.com/xing/xing-android-sdk/blob/master/sdk/src/test/java/com/xing/android/sdk/CompositeTypeJsonAdapterTest.java)

**Comment by serj-lotutovici** [comment]
I've made our custom type to implement `ParameterizedType`. But when I call `Types.newParameterizedType(List.class, myCustomType /* This is an instance. */)` we lose the custom implementation, since `ParameterizedTypeImpl` wraps every `ParameterizedType`.

Currently I found a workaround for this, and declaring my own ParameterizedType for lists.

Don't know if you guys what any actions on this, since adding functionality to avoid wrapping in `ParameterizedTypeImpl` is a loophole for bugs.
